cloud/aws/actuators/machine/actuator.go

```diff
@@ -141,21 +152,76 @@ func (a *Actuator) Delete(cluster *clusterv1.Cluster, machine *clusterv1.Machine
+// If the Update attempts to mutate any immutable state, the method will error
+// and no updates will be performed.
 func (a *Actuator) Update(cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
 	glog.Infof("Updating machine %v for cluster %v.", machine.Name, cluster.Name)
 
-	// Handling of machine config changes is not yet implemented.
-	// We should check which pieces of configuration have been updated, throw
-	// errors if an attempt is made to modify any immutable state, otherwise
-	// go ahead and modify what we can.
+	// Get the updated config. We're going to compare parts of this to the
+	// current Tags and Security Groups that AWS is aware of.
+	config, err := a.machineProviderConfig(machine.Spec.ProviderConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to get machine config")
+	}
 
 	// Get the new status from the provided machine object.
+	// We need this in case any of it has changed, and we also require the
+	// instanceID for various AWS queries.
 	status, err := a.machineProviderStatus(machine)
 	if err != nil {
 		return errors.Wrap(err, "failed to get machine status")
 	}
 
+	// Get the current instance description from AWS.
+	instanceDescription, err := a.ec2.InstanceIfExists(status.InstanceID)
+	if err != nil {
+		return errors.Wrap(err, "failed to get instance")
+	}
+
+	// We can now compare the various AWS state to the state we were passed.
+	// We will check immutable state first, in order to fail quickly before
+	// moving on to state that we can mutate.
+	// TODO: Implement immutable state check.
+
+	// Get the SecurityGroup annotations
+	securityGroupAnnotation, err := a.machineAnnotationJSON(machine, SecurityGroupsLastAppliedAnnotation)
+	if err != nil {
+		return errors.Wrap(err, "failed to get security group annotation")
+	}
+
+	// Now we check to see if we need to mutate any mutable state.
+	// Check if any additions have been made to the instance Security Groups.
+	sgChanged, sgIDs, sgAnnotation := a.securityGroupsChanged(
+		securityGroupAnnotation,
+		config.AdditionalSecurityGroups,
+		instanceDescription.SecurityGroups)
+	if sgChanged {
+		// Finally, update the instance with our new security groups.
+		err := a.ec2.UpdateNetworkInterfaceSecurityGroups(&instanceDescription.PrimaryInterface, sgIDs)
+		if err != nil {
+			return errors.Wrap(err, "failed to update network interface security groups")
+		}
+
+		// And then update the annotation to reflect our new managed state.
+		err = a.updateMachineAnnotationJSON(machine, SecurityGroupsLastAppliedAnnotation, sgAnnotation)
+		if err != nil {
+			return errors.Wrap(err, "failed to update security group annotation")
+		}
+	}
+
+	// Check if the instance tags were changed. If they were, update them.
+	// It would be possible here to only send new/updated tags, but for the
+	// moment we send everything, even if only a single tag was created or
+	// upated.
+	tagsChanged := a.tagsChanged(config.AdditionalTags, instanceDescription.Tags)
```

Comment:
Is the plan to update tag reconciliation with an annotation-based approach as well?

Reply:
Correct, however I have stopped on the tags part for the moment since there's another PR which looks like it's implementing a lot of tag helper methods. Once that is merged, I will resume working on the tags and use those new methods.